### ISSUE TITLE
Relationship between the weights of a model and the weights of its functional version

### ISSUE THREAD

**Issue by pratikgujjar** [facebookresearch/higher#59]
I am trying to implement the Meta-Pseudo Labels paper. (https://arxiv.org/pdf/2003.10580v2.pdf)
Which in summary is trying to train a student model (modelB) with a teacher model (modelA). Model A in turn tracks progress for the student by evaluating it on a validation set and then updates itself through a second order gradient update all the way back. So something like this using higher:

```python
modelA_optim = optim.Adam(modelA.parameters(), lr=0.01)
modelB_optim = optim.Adam(modelB.parameters(), lr=0.01)

for n in range(n_iter):
    with higher.innerloop_ctx(modelB, modelB_optim, copy_initial_weights=False) as (fmodel, diffopt):
        pseudo_labels = modelA(X_train)

        # Train model B on pseudo labels
        logits = fmodel(X_train)
        loss1 = CrossEntropy(logits, pseudo_labels)
        diffopt.step(loss1)

        # Update model A with model B's loss on a validation set
        logits_b = fmodel(X_val)
        loss2 = CrossEntropy(logits_b, y_val)

        # Also update model A with real labels
        logits_a = modelA(X_val)
        loss3 = CrossEntropy(logits_a, y_val)
        lossA = loss2+ loss3
        lossA.backward()

   modelA_optim.step()
   new_params = dict(fmodel.named_parameters())
   for name, params in modelB.named_parameters():
            params.data.copy_(new_params[name])
```

The last three lines of the code is where I have a problem. Because I want the fmodel's final values on say iteration 0, to be the starting point of the new fmodel at iteration 1, I explicitly copy the param values from fmodel to modelB at the end of each iteration. Although this runs without errors, when evaluated on a test set, the performance of this modelB is no better than random.

Does anyone know what I am doing wrong in here?

**Comment by mpariente** [comment]
Hey, have you tried loading with state dict, like so? 
```python
module.load_state_dict(fmodule.state_dict())
```

**Comment by mpariente** [comment]
Also, do you have `buffers` that change in you model? That could be a reason why this doesn't match.

**Comment by egrefen** [comment]
Hello @pratikgujjar. I vaguely see what you are trying to do, although without your full code it's hard to debug it myself (if I had time). There's a few things you should check, in addition to trying @mpariente's suggestions.

First, after computing `loss2`, is `list(torch.autograd.grad(loss2, modelA.parameters()))` a list of tensors, or are there `None`s in the list?

Second, if `ModelA` and `ModelB` are instances of the same class, after you've done your copy operation at the end of a loop iteration, can you do something like
```python
for p1, p2 in zip(modelA.parameters(), modelB.parameters()):
    print(torch.equal(p1, p2))  # should be all true if you copied correctly
```

As a separate point, I am a bit confused about this line
```python
pseudo_labels = modelA(X_train)
```
because later lines such as 
```python
logits_a = modelA(X_val)
```
seem to indicate that `modelA` returns logits, not the parameters of a categorical over the classes. Shouldn't you be passing these through a softmax before passing them to the cross-entropy with the predictions of `modelB`?

**Comment by pratikgujjar** [comment]
Thank you @mpariente and @egrefen for your comments. Loading state_dict fixes the issue for me. I suppose that when modelB was loaded with fmodel's weights by writing over its data_ attribute it didn't actually register. I say this because,

```
new_params = dict(fmodel.named_parameters())
for name, params in modelB.named_parameters():
        params.data.copy_(new_params[name])
```
^ results in bad modelB performance even though modelB and fmodel's parameter values are identical. But, 

```
new_params = dict(fmodel.named_parameters())
for name, params in modelB.named_parameters():
        params.data.copy_(new_params[name])
logits_b = modelB(X_val)
```
^ works just fine. I have no explanation for it and at the same time am afraid that I might not be doing a very good job explaining it here!

On @egrefen 's point, you're right! Loss1 is actually a KLDivergence whereas the other losses are all CrossEntropy. My bad!
